src/storage/storages.rs

```diff
@@ -332,3 +334,191 @@ impl<'a, C, T: UnprotectedStorage<C>> Join for &'a mut FlaggedStorage<C, T> {
+/// to *maintain* the cache and flagging correctly. 
+///
+///# Example Usage:
+/// 
+/// ```rust
+/// extern crate specs;
+/// use specs::prelude::*;
+/// 
+/// #[derive(PartialEq, Clone)]
+/// pub struct Comp(u32);
+/// impl Component for Comp {
+///     // `ChangedStorage` acts as a wrapper around another storage.
+///     // You can put any store inside of here (e.g. HashMapStorage, VecStorage, etc.)
+///     type Storage = ChangedStorage<Comp, VecStorage<Comp>>;
+/// }
+/// 
+/// pub struct CompSystem;
+/// impl<'a> System<'a> for CompSystem {
+///     type SystemData = WriteStorage<'a, Comp>;
+///     fn run(&mut self, mut comps: WriteStorage<'a, Comp>) {
+///         // Iterates over all components like normal.
+///         for comp in (&comps).join() {
+///             // ...
+///         }
+/// 
+///         // To iterate over the flagged/modified components:
+///         for flagged_comp in (&comps).open().1.join() {
+///             // ...
+///         }
+/// 
+///         // Modify the components in some way.
+///         for comp in (&mut comps).join() {
+///             // ...
+///         }
+///         // `maintain` checks for equality between the cache and the actual components 
+///         // and filters out the components that did not change from the flagged ones.
+///         (&mut comps).open().1.maintain();
+///
+///         // Clears the changed storage every frame with this system.
+///         (&mut comps).open().1.clear_flags();
+///     }
+/// }
+///# fn main() { }
+/// ```
+pub struct ChangedStorage<C, U> {
+    inner: FlaggedStorage<C, U>,
+    cache: U,
+    next: Option<Index>,
+}
+
+impl<C, U> ChangedStorage<C, U>
+    where C: PartialEq + Clone,
+          U: UnprotectedStorage<C>,
+{
+    /// Maintains the flagging and cache of the `ChangedStorage`.
+    /// 
+    /// Checks for equality between the cache and the actual components data
+    /// to filter out the components that did not change, leaving only the modified ones
+    /// in the flagged bitset.
+    ///
+    /// Should be called whenever you care about flagged components or alternatively
+    /// whenever the component storage is modified.
+    pub fn maintain(&mut self) {
+        if let Some(index) = self.next {
+            unsafe { 
+                if self.cache.get(index) == self.inner.get(index) {
+                    self.inner.unflag(index);
+                }
+                else {
+                    self.cache.insert(index, self.inner.get(index).clone());
+                }
+            }
+        }
+    }
+    // Forwarding methods to `FlaggedStorage`.
+    /// Whether the component related to the entity was flagged or not.
+    pub fn flagged<E: EntityIndex>(&self, entity: E) -> bool {
+        self.inner.flagged(entity)
+    }
+    /// Clears the bitset for flagged components.
+    ///
+    /// Should be called at least once depending on where you want to reset the flags.
```

Comment:
There are no right times to clear the flags, it depends on the context. If every system that sets these components to something clears them then the flags are completely useless to other systems for things like transforms. In other circumstances if you only want it to be flagged from a certain iteration then you can clear it before that iteration.